### PR TITLE
refactor(tfhe): create associated CompactPrivateKey and prepare casting

### PR DESCRIPTION
**Pull request by IceTDrinker**
closes: _please link all relevant issues_

### PR content/description

- for casting from the CompacthPublicKey parameter we need to add the notion of a kind on the CompactCiphertextList, where one kind will need to be cast thanks to an auxiliary keyswitching key and the other kind can just be expanded as before
- to avoid weird situations/corner cases we remove the ability to encrypt a "normal" ciphertext from a CompactPublicKey (which consisted in expanding right after encryption)

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/integer/ciphertext/compact_list.rs [hunk 1]
```diff
@@ -432,6 +436,10 @@ impl CompactCiphertextList {
         ))
     }
 
+    pub fn expand_with_casting(&self) {
+        todo!()
+    }
+
```
Comment (tmontaigu) [inline tfhe/src/integer/ciphertext/compact_list.rs:442]: Just to be sure, this is a todo for the next PR ?

File: tfhe/src/shortint/parameters/compact_public_key_only.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+        Ok(parameters)
+    }
+
+    pub const fn is_valid(&self) -> bool {
+        self.encryption_lwe_dimension.0.is_power_of_two()
+    }
+
+    /// This should be used while defining static parameters to verify they comply with the
+    /// requirements of ompact public key encryption.
```
Comment (tmontaigu) [inline tfhe/src/shortint/parameters/compact_public_key_only.rs:63]:
```suggestion
    /// requirements of compact public key encryption.
```

File: tfhe/src/integer/ciphertext/compact_list.rs (hunk 1 above)
Comment (IceTDrinker) [reply tfhe/src/integer/ciphertext/compact_list.rs:442]: I can remove it from here, at first I thought of doing everything at once but it seemed too much, so yes will be done in the next PR, I'll remove this to keep a consistent code base

**Review by tmontaigu — APPROVED** [review]
